src/api/slice.cc

```diff
@@ -0,0 +1,123 @@
+**/
+#include "slice.h"
+#include "../tainted/range.h"
+#include "../tainted/string_resource.h"
+#include "../tainted/transaction.h"
+#include "../iast.h"
+#include "v8.h"
+
+
+namespace iast {
+namespace api {
+
+using v8::FunctionCallbackInfo;
+using v8::Object;
+using v8::Local;
+using v8::Value;
+using v8::Isolate;
+
+using utils::GetLocalStringPointer;
+
+SharedRanges* getRangesInSlice(Transaction* transaction, SharedRanges* oRanges, int sliceStart, int sliceEnd) {
+    SharedRanges* newRanges = nullptr;
+    for (auto it = oRanges->begin(); it != oRanges->end(); ++it) {
+        auto oRange = *it;
+        int start, end;
+
+        if ((oRange->start < sliceStart) && (oRange->end <= sliceStart)) {
+            // range out of bounds (left)
+            continue;
+        }
+
+        if (oRange->start >= sliceEnd) {
+            // out of bounds (right), no need to keep iterating
+            break;
+        }
+
+        if ((oRange->start <= sliceStart) && (oRange->end > sliceEnd)) {
+            // range greater than slice
+            start = 0;
+            end = oRange->end - sliceStart;
+        } else if ((oRange->start >= sliceStart) && (oRange->end <= sliceEnd)) {
+            // range contained
+            start = oRange->start - sliceStart;
+            end = oRange->end - sliceStart;
+        } else if ((oRange->start < sliceStart) && (oRange->end <= sliceEnd)) {
+            // parcial left
+            start = 0;
+            end = oRange->end - sliceStart;
+        } else {
+            // parcial right
+            start = oRange->start - sliceStart;
+            end = sliceEnd;
+        }
+
+        if (!newRanges) {
+            newRanges = transaction->GetSharedVectorRange();
+        }
+
+        newRanges->PushBack(transaction->GetRange(start, end, oRange->inputInfo));
+    }
+    return newRanges;
+}
+
+void slice(const FunctionCallbackInfo<Value>& args) {
+    Isolate* isolate = args.GetIsolate();
+
+    if (args.Length() < 4) {
+        isolate->ThrowException(v8::Exception::TypeError(
+                        v8::String::NewFromUtf8(isolate,
+                        "Wrong number of arguments",
+                        v8::NewStringType::kNormal).ToLocalChecked()));
+        return;
+    }
+
+    auto context = isolate->GetCurrentContext();
+    auto vResult = args[1];
+    auto vSubject = args[2];
+    int sliceStart = args[3]->IntegerValue(context).FromJust();
+
+    args.GetReturnValue().Set(vResult);
+
+    Transaction* transaction = GetTransaction(GetLocalStringPointer(args[0]));
+    if (transaction == nullptr) {
+        return;
+    }
+
+    auto taintedObj = transaction->FindTaintedObject(GetLocalStringPointer(vSubject));
+    auto oRanges = taintedObj ? taintedObj->getRanges() : nullptr;
+
+    if (!oRanges) {
+        return;
+    }
+
+    try {
+        int subjectLength = v8::Local<v8::String>::Cast(vSubject)->Length();
```

Comment:
Can't we prevent doing this? It will be more optimal if we don't read this value, and we don't remove final ranges if `sliceEnd` is not defined

Reply:
Ok, I didn't realize that we need it when start is negative